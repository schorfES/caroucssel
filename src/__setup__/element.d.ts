export {};

declare global {
	interface Element {
		mockedTop?: number;
		mockedLeft?: number;
		mockedClientWidth: number;
		mockedOffsetWidth: number;
		mockedClientHeight: number;
		mockedOffsetHeight: number;
	}
}
