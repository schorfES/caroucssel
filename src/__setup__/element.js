function __getSiblings(el) {
	return [...el.parentNode.children];
}


HTMLElement.prototype.mockTop = null;
HTMLElement.prototype.mockLeft = null;
HTMLElement.prototype.mockClientWidth = 120;
HTMLElement.prototype.mockOffsetWidth = 120;
HTMLElement.prototype.mockClientHeight = 50;
HTMLElement.prototype.mockOffsetHeight = 50;

HTMLElement.prototype.scrollTo = function({top, left}) {
	this.mockTop = top;
	this.mockLeft = left;

	const event = document.createEvent('Event');
	event.initEvent('scroll');
	this.dispatchEvent(event);
	jest.runAllTimers();
};

HTMLElement.prototype.getBoundingClientRect = function() {
	const siblings = __getSiblings(this);
	const index = siblings.indexOf(this);
	const width = this.mockClientWidth;
	const height = this.mockClientHeight;
	const top = this.mockTop;
	const right = 0;
	const bottom = 0;
	let left = this.mockLeft;

	if (left === null) {
		left = index * width;
	}

	return {width, height, top, right, bottom, left};
};

Object.defineProperty(HTMLElement.prototype, 'scrollTop', {
	get: function() {
		return this.mockTop;
	}
});

Object.defineProperty(HTMLElement.prototype, 'scrollLeft', {
	get: function() {
		return this.mockLeft;
	}
});

Object.defineProperty(HTMLElement.prototype, 'clientWidth', {
	get: function() {
		return this.mockClientWidth;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
	get: function() {
		return this.mockOffsetWidth;
	}
});

Object.defineProperty(HTMLElement.prototype, 'clientHeight', {
	get: function() {
		return this.mockClientHeight;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
	get: function() {
		return this.mockOffsetHeight;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetTop', {
	get: function() {
		return this.getBoundingClientRect().top;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetLeft', {
	get: function() {
		return this.getBoundingClientRect().left;
	}
});
