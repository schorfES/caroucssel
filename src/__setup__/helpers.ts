/* istanbul ignore file */

/**
 * Options of the HTML markup.
 */
export type FixtureOptions = {
	id: string | null;
};

/**
 * Create a fixture of a basic carousel HTML markup with a specific amount of
 * items inside.
 * @internal
 * @param amount the number of items inside this HTML markup
 * @param options options to control the markup
 * @param options.id is the id of the carousel container element
 * @returns the rendered HTML markup
 */
export function fixture(amount: number, options: Partial<FixtureOptions> = {}): string {
	const settings: FixtureOptions = { id: null, ...options };

	return `
		<div class="container">
			<div class="caroucssel"${settings.id ? ` id="${settings.id}"` : ''}>
				${[...Array(amount).keys()].map((index) => `
					<div class="item item-${index}">Item ${index}</div>
				`).join('')}
			</div>
		</div>
	`;
}

/**
 * Triggers a resize event.
 * @internal
 */
export function triggerResize(): void {
	const event = document.createEvent('UIEvents');
	event.initEvent('resize', true, false);
	window.dispatchEvent(event);
	jest.runAllTimers();
}

/**
 * A scroll position.
 */
export type ScrollPosition = {
	left: number;
	top: number;
};

/**
 * Scrolls to a given position and triggers a scroll event on a given element.
 * @internal
 * @param element the element to control.
 * @param position new position of element
 */
export function triggerScroll(element: Element, position: Partial<ScrollPosition>): void {
	element.mockedTop = position.top || 0;
	element.mockedLeft = position.left || 0;

	const event = document.createEvent('Event');
	event.initEvent('scroll');
	element.dispatchEvent(event);
	jest.runAllTimers();
}

/**
 * Triggers a click event on a given element
 * @internal
 * @param element the element to control
 */
export function triggerClick(element: Element): void {
	const event = document.createEvent('Event');
	event.initEvent('click');
	element.dispatchEvent(event);
	jest.runAllTimers();
}

/**
 * Wrapper to ensure the querySelector() returns an element.
 * @internal
 * @param selector the selector passed into querySelector
 * @param root the element root from where to query
 * @returns the found element
 * @throws if no matching element was found
 */
export function querySelector(selector: string, root: Element | Document = document): Element {
	const el = root.querySelector(selector);
	if (!el) {
		throw new Error('Selector not found');
	}

	return el;
}
