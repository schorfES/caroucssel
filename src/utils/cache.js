const NAMESPACE = '__cache__';

export const fromCache = (instance, key, calculate) => {
	instance[NAMESPACE] = instance[NAMESPACE] || {};
	if (key in instance[NAMESPACE]) {
		return instance[NAMESPACE][key];
	}

	const value = calculate();
	instance[NAMESPACE][key] = value;
	return value;
};

export const clearCache = (instance, key) => {
	const cache = instance[NAMESPACE];
	if (!cache) {
		return;
	}

	delete(cache[key]);
};

export const clearFullCache = (instance) => delete(instance[NAMESPACE]);
