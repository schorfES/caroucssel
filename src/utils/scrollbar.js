export class Scrollbar {

	constructor() {
		this._d = null;
		window.addEventListener('resize', () => this._d = null);
	}

	get dimensions() {
		this._d = this._d || (() => {
			const
				inner = document.createElement('div'),
				outer = document.createElement('div')
			;
			let
				width, w1, w2,
				height, h1, h2
			;

			document.body.appendChild(outer);
			outer.style.position = 'absolute';
			outer.style.top = '0px';
			outer.style.left = '0px';
			outer.style.visibility = 'hidden';
			outer.appendChild(inner);

			// Calculate width:
			inner.style.width = '100%';
			inner.style.height = '200px';
			outer.style.width = '200px';
			outer.style.height = '150px';
			outer.style.overflow = 'hidden';
			w1 = inner.offsetWidth;
			outer.style.overflow = 'scroll';
			w2 = inner.offsetWidth;
			w2 = (w1 === w2) ? outer.clientWidth : w2;
			width = w1 - w2;

			// Calculate height:
			inner.style.width = '200px';
			inner.style.height = '100%';
			outer.style.width = '150px';
			outer.style.height = '200px';
			outer.style.overflow = 'hidden';
			h1 = inner.offsetHeight;
			outer.style.overflow = 'scroll';
			h2 = inner.offsetHeight;
			h2 = (h1 === h2) ? outer.clientHeight : h2;
			height = h1 - h2;

			document.body.removeChild(outer);
			return {width, height};
		})();

		return this._d;
	}

}
