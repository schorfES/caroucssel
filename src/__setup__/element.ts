import './element.d';


Object.defineProperties(Element.prototype, {
	mockedTop: {
		writable: true,
		value: undefined,
	},

	mockedLeft: {
		writable: true,
		value: undefined,
	},

	mockedClientWidth: {
		writable: true,
		value: 120,
	},

	mockedOffsetWidth: {
		writable: true,
		value: 120,
	},

	mockedClientHeight: {
		writable: true,
		value: 50,
	},

	mockedOffsetHeight: {
		writable: true,
		value: 50,
	},

	scrollTo: {
		writable: true,
		value: function(this: Element, { top, left }: { top: number, left: number }): void {
			this.mockedTop = top;
			this.mockedLeft = left;

			const event = document.createEvent('Event');
			event.initEvent('scroll');
			this.dispatchEvent(event);
			jest.runAllTimers();
		},
	},

	scrollTop: {
		get: function(this: Element) {
			return this.mockedTop;
		},
	},

	scrollLeft: {
		get: function(this: Element) {
			return this.mockedLeft;
		},
		set: function(this: Element, value: number) {
			this.mockedLeft = value;
		},
	},

	clientWidth: {
		get: function(this: Element) {
			return this.mockedClientWidth;
		},
	},

	clientHeight: {
		get: function(this: Element) {
			return this.mockedClientHeight;
		},
	},

	getBoundingClientRect: {
		writable: true,
		value: function(this: Element) {
			const siblings = [...this.parentNode?.children || []];
			const index = siblings.indexOf(this);

			return {
				x: 0,
				y: 0,
				width: this.mockedClientWidth,
				height: this.mockedClientHeight,
				top: this.mockedTop,
				right: 0,
				bottom: 0,
				left: this.mockedLeft ?? index * this.mockedClientWidth,
			};
		},
	},
});

Object.defineProperties(HTMLElement.prototype, {
	offsetWidth: {
		get: function(this: HTMLElement) {
			return this.mockedOffsetWidth;
		},
	},

	offsetHeight: {
		get: function(this: HTMLElement) {
			return this.mockedOffsetHeight;
		},
	},

	offsetTop: {
		get: function(this: HTMLElement) {
			return this.getBoundingClientRect().top;
		},
	},

	offsetLeft: {
		get: function(this: HTMLElement) {
			return this.getBoundingClientRect().left;
		},
	},
});
