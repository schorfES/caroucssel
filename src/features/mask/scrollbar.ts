import { clearCache, fromCache } from '../../utils/cache';


export type ScrollbarDimensions = {
	height: number,
};

/**
 * Helper class for scrollbar features.
 */
export class Scrollbar {

	/**
	 * Creates an instance.
	 * @internal
	 */
	constructor() {
		window.addEventListener('resize', () => {
			clearCache(this, 'dimensions');
		});
	}

	/**
	 * Calculates the dimensions of a scrollbar in the current browser. The result
	 * of the computation will be cached for this instance.
	 *
	 * Inspired by https://gist.github.com/kflorence/3086552
	 *
	 * @return the dimensions of the scrollar
	 */
	get dimensions(): ScrollbarDimensions {
		return fromCache<ScrollbarDimensions>(this, 'dimensions', () => {
			const inner = document.createElement('div');
			const outer = document.createElement('div');

			document.body.appendChild(outer);
			outer.style.position = 'absolute';
			outer.style.top = '0px';
			outer.style.left = '0px';
			outer.style.visibility = 'hidden';
			outer.appendChild(inner);

			// Disabled, not needed for current feature set.
			//
			// Calculate width:
			// inner.style.width = '100%';
			// inner.style.height = '200px';
			// outer.style.width = '200px';
			// outer.style.height = '150px';
			// outer.style.overflow = 'hidden';
			// w1 = inner.offsetWidth;
			// outer.style.overflow = 'scroll';
			// w2 = inner.offsetWidth;
			// w2 = (w1 === w2) ? outer.clientWidth : w2;
			// width = w1 - w2;

			// Calculate height:
			inner.style.width = '200px';
			inner.style.height = '100%';
			outer.style.width = '150px';
			outer.style.height = '200px';
			outer.style.overflow = 'hidden';

			const h1 = inner.offsetHeight;
			outer.style.overflow = 'scroll';
			let h2 = inner.offsetHeight;
			h2 = (h1 === h2) ? outer.clientHeight : h2;
			const height = h1 - h2;

			document.body.removeChild(outer);

			return {
				// width,
				height,
			};
		});
	}

}
