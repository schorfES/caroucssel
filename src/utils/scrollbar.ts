import {clearCache, fromCache} from './cache';

export type ScrollbarDimensions = {
	height: number,
};

export class Scrollbar {

	constructor() {
		window.addEventListener('resize', () => {
			clearCache(this, 'dimensions');
		});
	}

	// Inspired by https://gist.github.com/kflorence/3086552
	get dimensions(): ScrollbarDimensions {
		return fromCache<ScrollbarDimensions>(this, 'dimensions', () => {
			const
				inner = document.createElement('div'),
				outer = document.createElement('div')
			;
			let
				// width, w1, w2,
				height, h1, h2
			;

			document.body.appendChild(outer);
			outer.style.position = 'absolute';
			outer.style.top = '0px';
			outer.style.left = '0px';
			outer.style.visibility = 'hidden';
			outer.appendChild(inner);

			// Disabled, not needed for current feature set.
			//
			// Calculate width:
			// inner.style.width = '100%';
			// inner.style.height = '200px';
			// outer.style.width = '200px';
			// outer.style.height = '150px';
			// outer.style.overflow = 'hidden';
			// w1 = inner.offsetWidth;
			// outer.style.overflow = 'scroll';
			// w2 = inner.offsetWidth;
			// w2 = (w1 === w2) ? outer.clientWidth : w2;
			// width = w1 - w2;

			// Calculate height:
			inner.style.width = '200px';
			inner.style.height = '100%';
			outer.style.width = '150px';
			outer.style.height = '200px';
			outer.style.overflow = 'hidden';
			h1 = inner.offsetHeight;
			outer.style.overflow = 'scroll';
			h2 = inner.offsetHeight;
			h2 = (h1 === h2) ? outer.clientHeight : h2;
			height = h1 - h2;

			document.body.removeChild(outer);

			return {
				// width,
				height
			};
		});
	}

}
