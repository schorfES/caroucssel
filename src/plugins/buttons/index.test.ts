import { fixture, querySelector, triggerClick, triggerResize, triggerScroll } from '../../__setup__/helpers';
import { Carousel } from '../../carousel';
import { ScrollbarDimensions } from '../../utils/scrollbar';

import { Buttons, Params } from './index';


let mockScrollbarDimensions: ScrollbarDimensions | null = null;

jest.mock('../../utils/scrollbar', () => {
	return {
		Scrollbar: class {
			get dimensions() {
				return mockScrollbarDimensions;
			}
		},
	};
});

describe('Buttons plugin', () => {

	beforeEach(() => {
		jest.useFakeTimers();
		mockScrollbarDimensions = { height: 0 };
	});


	afterEach(() => {
		jest.clearAllTimers();
		document.body.innerHTML = '';
		Carousel.resetInstanceCount();
	});

	it('should add buttons', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');
		const options = {
			plugins: [new Buttons()],
		};
		new Carousel(el, options);

		expect(document.body.innerHTML).toMatchSnapshot();
	});

	it('should add buttons with custom options', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');
		const options = {
			plugins: [new Buttons({
				className: 'custom-button-class',
				previousClassName: 'custom-previous-button-class',
				previousLabel: 'Custom previous label',
				previousTitle: 'Custom previous title',
				nextClassName: 'custom-next-button-class',
				nextLabel: 'Custom next label',
				nextTitle: 'Custom next title',
			})],
		};
		new Carousel(el, options);

		expect(document.body.innerHTML).toMatchSnapshot();
	});

	it('should add buttons with custom template', () => {
		document.body.innerHTML = fixture(3, { id: 'custom-id' });
		const el = querySelector('.caroucssel');
		const options = {
			template: jest.fn<string, Params[]>(({ className, label, title }) =>
				`<span class="${className}" title="${title}">${label}</span>`),
			className: 'custom-button-class',

			nextClassName: 'custom-next-button-class',
			nextLabel: 'Custom next label',
			nextTitle: 'Custom next title',

			previousClassName: 'custom-previous-button-class',
			previousLabel: 'Custom previous label',
			previousTitle: 'Custom previous title',
		};
		new Carousel(el, {
			plugins: [new Buttons(options)],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(options.template).toHaveBeenCalledTimes(2);
		expect(options.template).toHaveBeenNthCalledWith(1, {
			className: 'custom-button-class custom-next-button-class',
			controls: 'custom-id',
			label: 'Custom next label',
			title: 'Custom next title',
		});
		expect(options.template).toHaveBeenNthCalledWith(2, {
			className: 'custom-button-class custom-previous-button-class',
			controls: 'custom-id',
			label: 'Custom previous label',
			title: 'Custom previous title',
		});
	});

	it('should handle buttons with custom template that returns empty string', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		const template: jest.Mock<string, []> = jest.fn(() => '');

		new Carousel(el, {
			plugins: [new Buttons({ template })],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(2);
	});

	it('should handle buttons with custom template that returns null', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		// Test when js custom implementation returns null
		// @ts-ignore
		const template: jest.Mock<string, []> = jest.fn(() => null);

		new Carousel(el, {
			plugins: [new Buttons({ template })],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(2);
	});

	it('should handle buttons with custom template that returns undefined', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		// Test when js custom implementation returns undefined
		// @ts-ignore
		const template: jest.Mock<string, [Params]> = jest.fn(() => undefined);

		new Carousel(el, {
			plugins: [new Buttons({ template })],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(2);
	});


	it('should add disabled buttons without items', () => {
		document.body.innerHTML = fixture(0);
		const el = querySelector('.caroucssel');
		const options = {
			plugins: [new Buttons()],
		};
		new Carousel(el, options);

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(document.querySelectorAll('.button[disabled]')).toHaveLength(2);
	});

	it('should re-render buttons on resize', () => {
		document.body.innerHTML = fixture(4);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		const options = {
			plugins: [new Buttons()],
		};
		new Carousel(el, options);

		expect(document.querySelectorAll('.button[disabled]')).toHaveLength(1);

		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 25);
		triggerResize();
		expect(document.querySelectorAll('.button[disabled]')).toHaveLength(2);
	});

	it('should update buttons on scroll', () => {
		document.body.innerHTML = fixture(6);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		const options = {
			plugins: [new Buttons()],
		};
		new Carousel(el, options);

		const buttons = document.querySelectorAll<HTMLButtonElement>('.button');
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([true, false]);

		triggerScroll(el, { left: 100 });
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, false]);

		triggerScroll(el, { left: 200 });
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, true]);
	});

	it('should update buttons manually', () => {
		document.body.innerHTML = fixture(6);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		const options = {
			plugins: [new Buttons()],
		};
		const carousel = new Carousel(el, options);

		const buttons = document.querySelectorAll<HTMLButtonElement>('.button');
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([true, false]);

		el.scrollTo({ left: 100 });
		carousel.update();
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, false]);

		el.scrollTo({ left: 200 });
		carousel.update();
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, true]);
	});

	it('should handle clicks', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 100);
		const options = {
			plugins: [new Buttons()],
		};
		new Carousel(el, options);

		const callback = jest.spyOn(el, 'scrollTo');
		const buttons = document.querySelectorAll<HTMLButtonElement>('.button');

		triggerClick(buttons[1]); // navigate forwards
		expect(callback).toHaveBeenCalledTimes(1);
		expect(callback).toHaveBeenNthCalledWith(1, { left: 100, behavior: 'smooth' });
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, false]);

		triggerClick(buttons[1]); // navigate forwards
		expect(callback).toHaveBeenCalledTimes(2);
		expect(callback).toHaveBeenNthCalledWith(2, { left: 200, behavior: 'smooth' });
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, true]);

		triggerClick(buttons[1]); // navigate forwards (doesn't work, disabled)
		expect(callback).toHaveBeenCalledTimes(2);

		triggerClick(buttons[0]); // navigate backwards
		expect(callback).toHaveBeenCalledTimes(3);
		expect(callback).toHaveBeenNthCalledWith(3, { left: 100, behavior: 'smooth' });
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([false, false]);

		triggerClick(buttons[0]); // navigate backwards
		expect(callback).toHaveBeenCalledTimes(4);
		expect(callback).toHaveBeenNthCalledWith(4, { left: 0, behavior: 'smooth' });
		expect([...buttons].map(({ disabled }) => disabled)).toEqual([true, false]);

		triggerClick(buttons[0]); // navigate backwards (doesn't work, disabled)
		expect(callback).toHaveBeenCalledTimes(4);
	});

});
