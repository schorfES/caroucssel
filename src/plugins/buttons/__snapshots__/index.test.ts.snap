// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons plugin should add buttons 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
  <button type="button"
          class="button is-previous"
          aria-label="Previous"
          title="Go to previous"
          aria-controls="caroucssel-1"
          disabled
  >
    <span>
      Previous
    </span>
  </button>
  <button type="button"
          class="button is-next"
          aria-label="Next"
          title="Go to next"
          aria-controls="caroucssel-1"
  >
    <span>
      Next
    </span>
  </button>
</div>
`;

exports[`Buttons plugin should add buttons with custom options 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
  <button type="button"
          class="custom-button-class custom-previous-button-class"
          aria-label="Custom previous label"
          title="Custom previous title"
          aria-controls="caroucssel-1"
          disabled
  >
    <span>
      Custom previous label
    </span>
  </button>
  <button type="button"
          class="custom-button-class custom-next-button-class"
          aria-label="Custom next label"
          title="Custom next title"
          aria-controls="caroucssel-1"
  >
    <span>
      Custom next label
    </span>
  </button>
</div>
`;

exports[`Buttons plugin should add buttons with custom template 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="custom-id"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
  <span class="custom-button-class custom-previous-button-class"
        title="Custom previous title"
  >
    Custom previous label
  </span>
  <span class="custom-button-class custom-next-button-class"
        title="Custom next title"
  >
    Custom next label
  </span>
</div>
`;

exports[`Buttons plugin should add disabled buttons without items 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
    </div>
  </div>
  <button type="button"
          class="button is-previous"
          aria-label="Previous"
          title="Go to previous"
          aria-controls="caroucssel-1"
          disabled
  >
    <span>
      Previous
    </span>
  </button>
  <button type="button"
          class="button is-next"
          aria-label="Next"
          title="Go to next"
          aria-controls="caroucssel-1"
          disabled
  >
    <span>
      Next
    </span>
  </button>
</div>
`;

exports[`Buttons plugin should handle buttons with custom template that returns empty string 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
</div>
`;

exports[`Buttons plugin should handle buttons with custom template that returns null 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
</div>
`;

exports[`Buttons plugin should handle buttons with custom template that returns undefined 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
</div>
`;
